<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="description" content="">
    <meta name="author" content="">
    <title></title>

    <!-- Le styles -->
    <link href="bootstrap/css/bootstrap.css" rel="stylesheet">
    <link href="bootstrap/css/bootstrap-responsive.css" rel="stylesheet">
    <link href="bootstrap/css/docs.css" rel="stylesheet">
    <link href="bootstrap/js/google-code-prettify/prettify.css" rel="stylesheet">

  </head>

  <body data-spy="scroll" data-target=".bs-docs-sidebar">

  

  <div class="container">

  <div class="row">

<div class="span3 bs-docs-sidebar"><ul class="nav nav-list bs-docs-sidenav" data-spy="affix"></ul></div><div class="span9"><!-- type=misc -->

<p>Any time an <code>Error</code> object is routed through a domain, a few extra fields
are added to it.</p>

<ul>
<li><code>error.domain</code> The domain that first handled the error.</li>
<li><code>error.domainEmitter</code> The event emitter that emitted an <code>'error'</code> event
with the error object.</li>
<li><code>error.domainBound</code> The callback function which was bound to the
domain, and passed an error as its first argument.</li>
<li><code>error.domainThrown</code> A boolean indicating whether the error was
thrown, emitted, or passed to a bound callback function.</li>
</ul>

<p>The Domain class encapsulates the functionality of routing errors and
uncaught exceptions to the active Domain object.</p>

<p>Domain is a child class of [<code>EventEmitter</code>][].  To handle the errors that it
catches, listen to its <code>'error'</code> event.</p>

<blockquote>
  <p>稳定性: 0 - 废弃的</p>
</blockquote>

<p><strong>该模块即将废弃</strong>。Once a replacement API has been
finalized, this module will be fully deprecated. Most end users should
<strong>not</strong> have cause to use this module. Users who absolutely must have
the functionality that domains provide may rely on it for the time being
but should expect to have to migrate to a different solution
in the future.</p>

<p>Domains provide a way to handle multiple different IO operations as a
single group.  If any of the event emitters or callbacks registered to a
domain emit an <code>'error'</code> event, or throw an error, then the domain object
will be notified, rather than losing the context of the error in the
<code>process.on('uncaughtException')</code> handler, or causing the program to
exit immediately with an error code.</p>

<ul>
<li><code>emitter</code> {EventEmitter|Timer} emitter or timer to be added to the domain</li>
</ul>

<p>Explicitly adds an emitter to the domain.  If any event handlers called by
the emitter throw an error, or if the emitter emits an <code>'error'</code> event, it
will be routed to the domain's <code>'error'</code> event, just like with implicit
binding.</p>

<p>This also works with timers that are returned from [<code>setInterval()</code>][] and
[<code>setTimeout()</code>][].  If their callback function throws, it will be caught by
the domain 'error' handler.</p>

<p>If the Timer or EventEmitter was already bound to a domain, it is removed
from that one, and bound to this one instead.</p>

<ul>
<li><code>callback</code> {Function} The callback function</li>
<li>Returns: {Function} The bound function</li>
</ul>

<p>The returned function will be a wrapper around the supplied callback
function.  When the returned function is called, any errors that are
thrown will be routed to the domain's <code>'error'</code> event.</p>

<ul>
<li>Returns: {Domain}</li>
</ul>

<p>Returns a new Domain object.</p>

<blockquote>
  <p>Stability: 0 - Deprecated.  Please recover from failed IO actions
  explicitly via error event handlers set on the domain.</p>
</blockquote>

<p>Once <code>dispose</code> has been called, the domain will no longer be used by callbacks
bound into the domain via <code>run</code>, <code>bind</code>, or <code>intercept</code>, and a <code>'dispose'</code> event
is emitted.</p>

<p>The <code>enter</code> method is plumbing used by the <code>run</code>, <code>bind</code>, and <code>intercept</code>
methods to set the active domain. It sets <code>domain.active</code> and <code>process.domain</code>
to the domain, and implicitly pushes the domain onto the domain stack managed
by the domain module (see [<code>domain.exit()</code>][] for details on the domain stack). The
call to <code>enter</code> delimits the beginning of a chain of asynchronous calls and I/O
operations bound to a domain.</p>

<p>Calling <code>enter</code> changes only the active domain, and does not alter the domain
itself. <code>enter</code> and <code>exit</code> can be called an arbitrary number of times on a
single domain.</p>

<p>If the domain on which <code>enter</code> is called has been disposed, <code>enter</code> will return
without setting the domain.</p>

<p>The <code>exit</code> method exits the current domain, popping it off the domain stack.
Any time execution is going to switch to the context of a different chain of
asynchronous calls, it's important to ensure that the current domain is exited.
The call to <code>exit</code> delimits either the end of or an interruption to the chain
of asynchronous calls and I/O operations bound to a domain.</p>

<p>If there are multiple, nested domains bound to the current execution context,
<code>exit</code> will exit any domains nested within this domain.</p>

<p>Calling <code>exit</code> changes only the active domain, and does not alter the domain
itself. <code>enter</code> and <code>exit</code> can be called an arbitrary number of times on a
single domain.</p>

<p>If the domain on which <code>exit</code> is called has been disposed, <code>exit</code> will return
without exiting the domain.</p>

<ul>
<li><code>callback</code> {Function} The callback function</li>
<li>Returns: {Function} The intercepted function</li>
</ul>

<p>This method is almost identical to [<code>domain.bind(callback)</code>][].  However, in
addition to catching thrown errors, it will also intercept [<code>Error</code>][]
objects sent as the first argument to the function.</p>

<p>In this way, the common <code>if (err) return callback(err);</code> pattern can be replaced
with a single error handler in a single place.</p>

<ul>
<li>{Array}</li>
</ul>

<p>An array of timers and event emitters that have been explicitly added
to the domain.</p>

<ul>
<li><code>emitter</code> {EventEmitter|Timer} emitter or timer to be removed from the domain</li>
</ul>

<p>The opposite of [<code>domain.add(emitter)</code>][].  Removes domain handling from the
specified emitter.</p>

<ul>
<li><code>fn</code> {Function}</li>
<li><code>...args</code> {any}</li>
</ul>

<p>Run the supplied function in the context of the domain, implicitly
binding all event emitters, timers, and lowlevel requests that are
created in that context. Optionally, arguments can be passed to
the function.</p>

<p>This is the most basic way to use a domain.</p>

<p>Example:</p>

<pre class="prettyprint"><code>const domain = require('domain');
const fs = require('fs');
const d = domain.create();
d.on('error', (er) =&gt; {
  console.error('Caught error!', er);
});
d.run(() =&gt; {
  process.nextTick(() =&gt; {
    setTimeout(() =&gt; { // simulating some various async stuff
      fs.open('non-existent file', 'r', (er, fd) =&gt; {
        if (er) throw er;
        // proceed...
      });
    }, 100);
  });
});
</code></pre>

<p>In this example, the <code>d.on('error')</code> handler will be triggered, rather
than crashing the program.</p>

<pre class="prettyprint"><code>const d = domain.create();

function readSomeFile(filename, cb) {
  fs.readFile(filename, 'utf8', d.bind((er, data) =&gt; {
    // if this throws, it will also be passed to the domain
    return cb(er, data ? JSON.parse(data) : null);
  }));
}

d.on('error', (er) =&gt; {
  // an error occurred somewhere.
  // if we throw it now, it will crash the program
  // with the normal line number and stack message.
});





const d = domain.create();

function readSomeFile(filename, cb) {
  fs.readFile(filename, 'utf8', d.intercept((data) =&gt; {
    // note, the first argument is never passed to the
    // callback since it is assumed to be the 'Error' argument
    // and thus intercepted by the domain.

    // if this throws, it will also be passed to the domain
    // so the error-handling logic can be moved to the 'error'
    // event on the domain instead of being repeated throughout
    // the program.
    return cb(null, JSON.parse(data));
  }));
}

d.on('error', (er) =&gt; {
  // an error occurred somewhere.
  // if we throw it now, it will crash the program
  // with the normal line number and stack message.
});
</code></pre>

<!--type=misc-->

<p>Sometimes, the domain in use is not the one that ought to be used for a
specific event emitter.  Or, the event emitter could have been created
in the context of one domain, but ought to instead be bound to some
other domain.</p>

<p>For example, there could be one domain in use for an HTTP server, but
perhaps we would like to have a separate domain to use for each request.</p>

<p>That is possible via explicit binding.</p>

<p>For example:</p>

<pre class="prettyprint"><code>// create a top-level domain for the server
const domain = require('domain');
const http = require('http');
const serverDomain = domain.create();

serverDomain.run(() =&gt; {
  // server is created in the scope of serverDomain
  http.createServer((req, res) =&gt; {
    // req and res are also created in the scope of serverDomain
    // however, we'd prefer to have a separate domain for each request.
    // create it first thing, and add req and res to it.
    const reqd = domain.create();
    reqd.add(req);
    reqd.add(res);
    reqd.on('error', (er) =&gt; {
      console.error('Error', er, req.url);
      try {
        res.writeHead(500);
        res.end('Error occurred, sorry.');
      } catch (er2) {
        console.error('Error sending 500', er2, req.url);
      }
    });
  }).listen(1337);
});
</code></pre>

<!--type=misc-->

<p>If domains are in use, then all <strong>new</strong> EventEmitter objects (including
Stream objects, requests, responses, etc.) will be implicitly bound to
the active domain at the time of their creation.</p>

<p>Additionally, callbacks passed to lowlevel event loop requests (such as
to fs.open, or other callback-taking methods) will automatically be
bound to the active domain.  If they throw, then the domain will catch
the error.</p>

<p>In order to prevent excessive memory usage, Domain objects themselves
are not implicitly added as children of the active domain.  If they
were, then it would be too easy to prevent request and response objects
from being properly garbage collected.</p>

<p>If you <em>want</em> to nest Domain objects as children of a parent Domain,
then you must explicitly add them.</p>

<p>Implicit binding routes thrown errors and <code>'error'</code> events to the
Domain's <code>'error'</code> event, but does not register the EventEmitter on the
Domain, so [<code>domain.dispose()</code>][] will not shut down the EventEmitter.
Implicit binding only takes care of thrown errors and <code>'error'</code> events.</p>

<!-- type=misc -->

<p>Domain error handlers are not a substitute for closing down your
process when an error occurs.</p>

<p>By the very nature of how [<code>throw</code>][] works in JavaScript, there is almost
never any way to safely "pick up where you left off", without leaking
references, or creating some other sort of undefined brittle state.</p>

<p>The safest way to respond to a thrown error is to shut down the
process.  Of course, in a normal web server, you might have many
connections open, and it is not reasonable to abruptly shut those down
because an error was triggered by someone else.</p>

<p>The better approach is to send an error response to the request that
triggered the error, while letting the others finish in their normal
time, and stop listening for new requests in that worker.</p>

<p>In this way, <code>domain</code> usage goes hand-in-hand with the cluster module,
since the master process can fork a new worker when a worker
encounters an error.  For Node.js programs that scale to multiple
machines, the terminating proxy or service registry can take note of
the failure, and react accordingly.</p>

<p>For example, this is not a good idea:</p>

<pre class="prettyprint"><code>// XXX WARNING!  BAD IDEA!

const d = require('domain').create();
d.on('error', (er) =&gt; {
  // The error won't crash the process, but what it does is worse!
  // Though we've prevented abrupt process restarting, we are leaking
  // resources like crazy if this ever happens.
  // This is no better than process.on('uncaughtException')!
  console.log(`error, but oh well ${er.message}`);
});
d.run(() =&gt; {
  require('http').createServer((req, res) =&gt; {
    handleRequest(req, res);
  }).listen(PORT);
});
</code></pre>

<p>By using the context of a domain, and the resilience of separating our
program into multiple worker processes, we can react more
appropriately, and handle errors with much greater safety.</p>

<pre class="prettyprint"><code>// Much better!

const cluster = require('cluster');
const PORT = +process.env.PORT || 1337;

if (cluster.isMaster) {
  // In real life, you'd probably use more than just 2 workers,
  // and perhaps not put the master and worker in the same file.
  //
  // You can also of course get a bit fancier about logging, and
  // implement whatever custom logic you need to prevent DoS
  // attacks and other bad behavior.
  //
  // See the options in the cluster documentation.
  //
  // The important thing is that the master does very little,
  // increasing our resilience to unexpected errors.

  cluster.fork();
  cluster.fork();

  cluster.on('disconnect', (worker) =&gt; {
    console.error('disconnect!');
    cluster.fork();
  });

} else {
  // the worker
  //
  // This is where we put our bugs!

  const domain = require('domain');

  // See the cluster documentation for more details about using
  // worker processes to serve requests.  How it works, caveats, etc.

  const server = require('http').createServer((req, res) =&gt; {
    const d = domain.create();
    d.on('error', (er) =&gt; {
      console.error(`error ${er.stack}`);

      // Note: we're in dangerous territory!
      // By definition, something unexpected occurred,
      // which we probably didn't want.
      // Anything can happen now!  Be very careful!

      try {
        // make sure we close down within 30 seconds
        const killtimer = setTimeout(() =&gt; {
          process.exit(1);
        }, 30000);
        // But don't keep the process open just for that!
        killtimer.unref();

        // stop taking new requests.
        server.close();

        // Let the master know we're dead.  This will trigger a
        // 'disconnect' in the cluster master, and then it will fork
        // a new worker.
        cluster.worker.disconnect();

        // try to send an error to the request that triggered the problem
        res.statusCode = 500;
        res.setHeader('content-type', 'text/plain');
        res.end('Oops, there was a problem!\n');
      } catch (er2) {
        // oh well, not much we can do at this point.
        console.error(`Error sending 500! ${er2.stack}`);
      }
    });

    // Because req and res were created before this domain existed,
    // we need to explicitly add them.
    // See the explanation of implicit vs explicit binding below.
    d.add(req);
    d.add(res);

    // Now run the handler function in the domain.
    d.run(() =&gt; {
      handleRequest(req, res);
    });
  });
  server.listen(PORT);
}

// This part is not important.  Just an example routing thing.
// You'd put your fancy application logic here.
function handleRequest(req, res) {
  switch (req.url) {
    case '/error':
      // We do some async stuff, and then...
      setTimeout(() =&gt; {
        // Whoops!
        flerb.bark();
      });
      break;
    default:
      res.end('ok');
  }
}
</code></pre>
      </div>
    </div>

  </div>

    <!-- Le javascript
    ================================================== -->
    <!-- Placed at the end of the document so the pages load faster -->
    <script src="bootstrap/js/jquery.js"></script>
    <script src="bootstrap/js/google-code-prettify/prettify.js"></script>
    <script src="bootstrap/js/bootstrap.js"></script>
    <script type="text/javascript">prettyPrint();</script>

  </body>
</html>
